XXX-asyncwrap-api.md

```diff
@@ -0,0 +1,309 @@
+| Title  | AsyncWrap API |
+|--------|---------------|
+| Author | @trevnorris   |
+| Status | DRAFT         |
+| Date   | 2016-04-05    |
+
+## Description
+
+Since its initial introduction with the `AsyncListener` API, `AsyncWrap` has
+slowly evolved to ensure a generalized API that would serve as a solid base for
+module authors who wished to add events to the life cycle of the event loop.
+The API presented below aims to accomplish this.
+
+## API
+
+### Overview
+
+```js
+// Standard way of requiring a module. Snake case follows core module
+// convention.
+const async_wrap = require('async_wrap');
+
+// List of asynchronous providers (e.g. CRYPTO or TCPWRAP). Can be used for
+// trace filtering of events.
+const providers = async_wrap.Providers;
+
+// init() is called during object construction. The object handle being
+// constructed can be accessed as "this". Because init() is called during
+// construction it is not safe to blindly inspect the constructed object at the
+// moment. Though we are currently investigating how to make this safe.
+function init(id, type, parentId) { }
+
+// pre() is called just before the handle's callback is called. It can be
+// called 0-N times for handles (e.g. TCPWRAP), and should be called exactly 1
+// time for requests (e.g. FSREQWRAP).
+function pre(id) { }
+
+// post() is called just after the handle's callback has finished, and will be
+// called regardless whether the handle's callback threw. If the handle's
+// callback did throw then hasThrown will be true.
+function post(id, hasThrown) { }
+
+// destroy() is called when an AsyncWrap instance is destroyed. In cases like
+// HTTPPARSER where the resource is reused, or timers where the handle is only
+// a JS object, destroy() will be triggered manually soon after post() has
+// completed. This is the only callback that does not receive the context as
+// "this". Because it is called while the associated class is being destructed
+// it is unsafe to operate on the object.
+function destroy(id) { }
+
+// Add a set of hooks to be called during the lifetime of asynchronous events.
+// createHook() returns an instance of the AsyncHook constructor that can
+// control how those hooks are used. For example, enabling or disabling their
+// execution.
+asyncHook = async_wrap.createHook({ init, pre, post, destroy });
+
+// Allow callbacks of this AsyncHook instance to fire. This is not an implicit
+// action after running createHook(), and must be explicitly run to being
+// executing callbacks.
+asyncHook.enable();
```

Comment:
@mhdawson That is coming in the future, but not spec'd for initial API. It would roughly look something like (mind you I'm changing API names on the fly, but you'll get what I mean):
```js
const AsyncHook = require('async_wrap');
const pflags = AsyncHook.ProviderFlags;  // Providers by flag enum
const ah = new AsyncHook({ init, pre, post });
ah.filter(pflags.TCPWRAP | pflags.UDPWRAP | pflags.CRYPTO);
ah.enable();
```
The above will only trigger the callbacks for the three flags provided. There are a couple things ATM that need to be addressed:
1) Currently stored in an enum, not by or-able flag. Easy fix. Though worried we may run out of room.
2) Is there some type of propagation that should happen automatically under the hood? Otherwise the user is left with no stack traces, resource tracing, etc. Because of this I'm considering how useful this feature would be.